File: utils/pagination/pagination.py
"""Pagination helper functions."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any

from utils.error_handling import BadRequestError

from .constants import DEFAULT_PAGE, DEFAULT_SIZE, MAX_SIZE, PAGE_KEY, SIZE_KEY
from .filters import apply_filters, apply_query_filters, parse_filters
from .response import page_response
from .sort import apply_query_sort, apply_sort, parse_sort


@dataclass(frozen=True)
class Page:
    """Parsed page settings."""

    page: int = DEFAULT_PAGE
    size: int = DEFAULT_SIZE

    @property
    def offset(self) -> int:
        """Return the row offset for this page."""
        return (self.page - 1) * self.size


def parse_page(params: Mapping[str, Any] | None) -> Page:
    """Parse and validate page/size values."""
    src = params or {}
    page = _int(src.get(PAGE_KEY, DEFAULT_PAGE), PAGE_KEY)
    size = _int(src.get(SIZE_KEY, DEFAULT_SIZE), SIZE_KEY)
    if page < 1:
        raise BadRequestError("page must be greater than zero")
    if size < 1 or size > MAX_SIZE:
        raise BadRequestError(f"size must be between 1 and {MAX_SIZE}")
    return Page(page=page, size=size)


def paginate_items(items: Sequence[Any], params: Mapping[str, Any] | None = None) -> dict[str, Any]:
    """Filter, sort, and paginate a Python sequence."""
    pg = parse_page(params)
    vals = list(items)
    vals = apply_filters(vals, parse_filters(params))
    vals = apply_sort(vals, parse_sort(params))
    total = len(vals)
    chunk = vals[pg.offset : pg.offset + pg.size]
    return page_response(chunk, page=pg.page, size=pg.size, total=total)


def apply_query(query: Any, model: Any, params: Mapping[str, Any] | None = None) -> Any:
    """Apply filters, sorts, limit, and offset to a SQLAlchemy-style query."""
    pg = parse_page(params)
    query = apply_query_filters(query, model, parse_filters(params))
    query = apply_query_sort(query, model, parse_sort(params))
    return query.offset(pg.offset).limit(pg.size)


def _int(val: Any, name: str) -> int:
    try:
        return int(val)
    except (TypeError, ValueError) as exc:
        raise BadRequestError(f"{name} must be an integer") from exc

File: utils/pagination/pagination.js
import { BadRequestError } from '../errorHandling.js';
import { ASC, DEFAULT_PAGE, DEFAULT_SIZE, DESC, FILTER_PREFIX, MAX_SIZE, OPS, ORDER_KEY, PAGE_KEY, SIZE_KEY, SORT_KEY } from './constants.js';
import { pageResponse } from './response.js';

export class Page {
  constructor({ page = DEFAULT_PAGE, size = DEFAULT_SIZE } = {}) { this.page = page; this.size = size; }
  get offset() { return (this.page - 1) * this.size; }
}
export class Filter { constructor({ field, op, val }) { this.field = field; this.op = op; this.val = val; } }
export class Sort { constructor({ field, desc = false }) { this.field = field; this.desc = desc; } }

export function parsePage(params = {}) {
  const page = parseInteger(params[PAGE_KEY] ?? DEFAULT_PAGE, PAGE_KEY);
  const size = parseInteger(params[SIZE_KEY] ?? DEFAULT_SIZE, SIZE_KEY);
  if (page < 1) throw new BadRequestError('page must be greater than zero');
  if (size < 1 || size > MAX_SIZE) throw new BadRequestError(`size must be between 1 and ${MAX_SIZE}`);
  return new Page({ page, size });
}

export function parseFilters(params = {}) {
  const items = [];
  for (const [key, val] of Object.entries(params ?? {})) {
    if (!key.startsWith(FILTER_PREFIX)) continue;
    const raw = key.slice(FILTER_PREFIX.length);
    const [field, opPart] = raw.split('__');
    const op = opPart || 'eq';
    if (!field || !OPS.has(op)) throw new BadRequestError('Invalid filter parameter.', { details: { filter: key } });
    items.push(new Filter({ field, op, val }));
  }
  return items;
}

export function parseSort(params = {}) {
  const raw = params?.[SORT_KEY];
  if (!raw) return [];
  const order = String(params[ORDER_KEY] ?? ASC).toLowerCase();
  if (![ASC, DESC].includes(order)) throw new BadRequestError('order must be asc or desc');
  const vals = Array.isArray(raw) ? raw : String(raw).split(',');
  return vals.filter((item) => String(item).trim()).map((item) => {
    const name = String(item).trim();
    const desc = name.startsWith('-') || order === DESC;
    const field = ['+', '-'].includes(name[0]) ? name.slice(1) : name;
    if (!field) throw new BadRequestError('Invalid sort field.');
    return new Sort({ field, desc });
  });
}

export function paginateItems(items, params = {}) {
  const pg = parsePage(params);
  let vals = applyFilters(items, parseFilters(params));
  vals = applySort(vals, parseSort(params));
  const total = vals.length;
  return pageResponse(vals.slice(pg.offset, pg.offset + pg.size), { page: pg.page, size: pg.size, total });
}

export function applyFilters(items, filters) {
  let vals = [...items];
  for (const filter of filters) vals = vals.filter((item) => match(valueOf(item, filter.field), filter.op, filter.val));
  return vals;
}

export function applySort(items, sorts) {
  const vals = [...items];
  [...sorts].reverse().forEach((rule) => {
    vals.sort((a, b) => compare(valueOf(a, rule.field), valueOf(b, rule.field)) * (rule.desc ? -1 : 1));
  });
  return vals;
}

export function applyQuery(query, model, params = {}) {
  return { query, model, page: parsePage(params), filters: parseFilters(params), sorts: parseSort(params) };
}

export function valueOf(item, field) { return item instanceof Map ? item.get(field) : item?.[field]; }
function parseInteger(val, name) { const parsed = Number.parseInt(val, 10); if (Number.isNaN(parsed)) throw new BadRequestError(`${name} must be an integer`); return parsed; }
function compare(left, right) { if (left === right) return 0; return left > right ? 1 : -1; }
function match(left, op, right) {
  if (op === 'eq') return left === right;
  if (op === 'ne') return left !== right;
  if (op === 'like') return String(left).toLowerCase().includes(String(right).toLowerCase());
  if (op === 'in') { const vals = Array.isArray(right) ? right : String(right).split(','); return vals.map(String).includes(String(left)); }
  if (op === 'lt') return left < right;
  if (op === 'lte') return left <= right;
  if (op === 'gt') return left > right;
  if (op === 'gte') return left >= right;
  return false;
}
